rfcs/0004-pablo-distribution.adoc

```diff
@@ -0,0 +1,480 @@
+
+
+
+[#_rewardaccumulationhook]
+==== [Needs modification] RewardAccumulationHook
+
+Following algorithm is to added as part of the existing https://github.com/ComposableFi/composable/blob/main/frame/staking-rewards/src/lib.rs#L363[block hook] in staking rewards pallet. As it is only accumulating new rewards for an upcoming epoch, the code is proposed to be run inside a new state `State:AccumulatingRewards`.
+
+[plantuml,images/staking-rewards-reward-accumulation-hook,png]
+----
+start
+->State:AccumulatingRewards;
+:rewardConfigIndex = read(rewardConfigCount) - 1;
+repeat :rewardConfig = read(rewardConfigs[rewardConfigIndex--]);
+    :reward_allocation = Assets::balance_of(reward_allocation_acc,
+            rewardConfig.asset);
+    :epochReward = reward_allocation * rewardConfig.reward_rate;
+    :stakerReward = epochReward;
+    :perAssetRewardAllocationIndex = read(perAssetRewardAllocationCount) - 1;
+    repeat :lpRewardAllocation = read(perAssetRewardAllocation[perAssetRewardAllocationIndex--]);
+        :lpEpochReward = EpochReward * lpRewardAllocation;
+        :staking_rewards::transfer_reward(
+            lpRewardAllocation.asset,
+            rewardConfig.asset,
+            reward_allocation_acc,
+            lpEpochReward, false);
+        :stakerReward -= lpEpochReward;
+    backward:perAssetRewardAllocationIndex;
+    repeat while (perAssetRewardAllocationIndex != -1)
+    :staking_rewards::transfer_reward(
+        rewardConfig.asset,
+        rewardConfig.asset,
+        reward_allocation_acc,
+        stakerReward, false);
+    note left: remainder after the per\nasset allocation is\ntransferred to same\nasset stakers reward
+backward:rewardConfigIndex;
+repeat while (rewardConfigIndex != -1)
+stop
+----
+
+This algorithm runs in `O(staked_asset_type_count * rewarded_asset_type_count)`.
+
+== Implementation
+
+=== Pallet Pablo: LP Fee + Staking Changes
+
+- [ ] Implement <<_feeconfig>> on pallet-pablo across all 3 types of pools.
+- [ ] Implement <<_pblo_staker_trading_fee_distribution>>.
+
+=== Pallet Staking Rewards: PICA/PBLO Staking Related Changes
+
+- [ ] Implement <<_rewardaccumulationhook>>.
+
+[appendix]
+== Fee Distribution Q&A
+
+Based on the current setup following questions arise when deciding on the distribution of these fees to relevant liquidity providers, owners and stakers.
+
+. A Protocol Fee for all pools in Pablo (or even protocol pallets other than Pablo)?
++
+Does it make sense to define a protocol fee percentage on top of the pool owner fees of the pools so that the protocol fee can be used as the pot out of which the stakers are rewarded? Initially the Protocol Fee = Pool Owner Fee as the pools are owned by Composable. Assumption here is that the stakers would indeed still get a reward out of third party created pool fees.
++
+*Comment:* While having a protocol funding mechanism is valuable, initially the protocol fees should zero or minimal.
```

Comment:
PICA is Coposable fee i guess. composable should motivate to use more pica. creating pools also motivatrs for PICA